Fix FEInterface for MG support

mwichro opened the pull request.

masterleinad commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -694,6 +724,38 @@ FEInterfaceValues<dim, spacedim>::at_boundary() const
+  Assert(
+    is_level_cell == false,
+    ExcMessage(
+      "get_dof_indices() only works on active cells. Use get_mg_interface_dof_indices instead."))
```
```suggestion
      "get_interface_dof_indices() only works on active cells. Use get_mg_interface_dof_indices instead."))
```

masterleinad commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -694,6 +724,38 @@ FEInterfaceValues<dim, spacedim>::at_boundary() const
+  Assert(
+    is_level_cell == true,
+    ExcMessage(
+      "get_mg_dof_indices() only works on mg cells. Use get_interface_dof_indices instead."))
```
```suggestion
      "get_mg_interface_dof_indices() only works on mg cells. Use get_interface_dof_indices instead."))
```

masterleinad commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -694,6 +724,38 @@ FEInterfaceValues<dim, spacedim>::at_boundary() const
+
+template <int dim, int spacedim>
+std::vector<types::global_dof_index>
+FEInterfaceValues<dim, spacedim>::get_active_or_mg_dof_indices() const
```
```suggestion
FEInterfaceValues<dim, spacedim>::get_active_or_mg_interface_dof_indices() const
```

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -235,6 +235,18 @@ class FEInterfaceValues
+  get_mg_interface_dof_indices() const;
+
+  /**
+   * Get dof indices for generic loops.
```
Can you expand what "generic loops" are? It's not a term we commonly use in deal.II (and in fact, I don't know what it means :-) )

bangerth commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -450,6 +462,10 @@ class FEInterfaceValues
+
+#ifdef DEBUG
+  bool is_level_cell;
+#endif
```
Let's not do this. We've tried to keep the debug and release libraries ABI compatible -- they may contain different code, but the sizes of structures and interfaces of functions should be the same. So just add the variable if you need it.

mwichro commented:
@tjhei  (and @masterleinad )
1) I have no idea how to write quick test for that. I am able to get MG  preconditioner and covergence rate seems to be constant. I've come up with the idea of smoother today, so it's a fresh thing. It also works in parallel: MPI, distributed triangulation.
2) I was trying to get the same interface as DoFCellAccessor where all functions :
get_active_or_mg_dof_indices 
get_dof_indices 
get_mg_dof_indices 
coexist. This also allows some assertions, so it may be easier to locate the bugs.

@bangerth  I'll update documentation and remove #ifdef DEBUG

masterleinad commented:
>     1. I have no idea how to write quick test for that.

Have a look at `tests/feinterface/fe_interface_values_01.cc`. Copy and modify accordingly.

tjhei commented:
> 2\. as trying to get the same interface as DoFCellAccessor where all functions :

My vote would be to reuse ``get_interface_dof_indices``.

mwichro commented:
@tjhei  I've changed to _get_dof_indices_ since this was the that returned active dof indices previously. Now in case of level cell it returns mg dof indices.
 I've fixed my name and email, but I don't know how to fix it in the commit.

masterleinad commented:
The error message says
```
Running indentation test on Pull Request #9978
invalid author 'Aazaram' without firstname and lastname
hint: for possible solutions, consult the webpage:
      https://github.com/dealii/dealii/wiki/Indentation#commit-authorship
```
and on that page is another link describing how to change your username or email for older commits.

tjhei commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -450,6 +453,7 @@ class FEInterfaceValues
    * as determined in reinit().
    */
   FEFaceValuesBase<dim> *fe_face_values_neighbor;
+
```
remove this line please

tjhei commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -228,6 +228,9 @@ class FEInterfaceValues
 
   /**
    * Return the set of joint DoF indices. This includes indices from both cells.
+   * If reinit was called on active cell the set of active dof indices are returned,
```
```suggestion
   * If reinit was called with an active cell iterator, the indices are based on the active indices (returned by `DoFCellAccessor::get_dof_indices()` ),
```

tjhei commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -228,6 +228,9 @@ class FEInterfaceValues
   /**
    * Return the set of joint DoF indices. This includes indices from both cells.
+   * If reinit was called on active cell the set of active dof indices are returned,
+   * in case of level cell (that is, if is_level_cell() return true )
```
please update, this function is no longer here.

tjhei commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -542,10 +546,12 @@ FEInterfaceValues<dim, spacedim>::reinit(
-    cell_neighbor->get_dof_indices(v2);
+    cell_neighbor->get_active_or_mg_dof_indices(v2);
+
+
```
remove empty lines

tjhei commented on include/deal.II/fe/fe_interface_values.h:
```diff
@@ -695,7 +701,8 @@ template <int dim, int spacedim>
 {
-  return interface_dof_indices;
+
+    return interface_dof_indices;
```
please revert this change as well.

tjhei commented:
Thank you, that looks much better now! Would you mind squashing your commits into a single commit? I am happy to contribute a test-case so you can see how to do it.

luca-heltai commented:
Somehow tidy is complaining about something unrelated to this PR.

masterleinad commented:
> Somehow tidy is complaining about something unrelated to this PR.

Yes, we need to rebase to get rid of it. I am fine with ignoring it here.
